http-client/src/main/java/com/facebook/airlift/http/client/jetty/ConcurrentScheduler.java

```diff
@@ -1,61 +1,57 @@
-import java.util.concurrent.ScheduledThreadPoolExecutor;
-import java.util.concurrent.ThreadFactory;
-import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 
-import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
-// based on ScheduledExecutorScheduler
-class ConcurrentScheduler
+public class ConcurrentScheduler
         extends AbstractLifeCycle
         implements Scheduler
 {
+    private Optional<ConcurrentScheduledExecutor> concurrentScheduler;
+    private final int schedulerCount;
     private final int threadsPerScheduler;
-    private final ScheduledExecutorService[] schedulers;
-    private final ThreadFactory threadFactory;
+    private final String threadNameFormat;
 
-    ConcurrentScheduler(int schedulerCount, int threadsPerScheduler, String threadBaseName)
+    public ConcurrentScheduler(int schedulerCount, int threadsPerScheduler, String threadNameFormat)
     {
         checkArgument(schedulerCount > 0, "schedulerCount must be at least one");
-        this.schedulers = new ScheduledThreadPoolExecutor[schedulerCount];
         checkArgument(threadsPerScheduler > 0, "threadsPerScheduler must be at least one");
+        requireNonNull(threadNameFormat, "threadNameFormat is null");
+
+        this.schedulerCount = schedulerCount;
         this.threadsPerScheduler = threadsPerScheduler;
-        requireNonNull(threadBaseName, "threadBaseName is null");
-        threadFactory = daemonThreadsNamed(threadBaseName + "-timeout-%s");
+        this.threadNameFormat = requireNonNull(threadNameFormat, "threadNameFormat is null");
+        concurrentScheduler = Optional.empty();
     }
 
     @Override
     protected void doStart()
     {
-        for (int i = 0; i < schedulers.length; i++) {
-            ScheduledThreadPoolExecutor scheduledExecutorService = new ScheduledThreadPoolExecutor(threadsPerScheduler, threadFactory);
-            scheduledExecutorService.setRemoveOnCancelPolicy(true);
-            schedulers[i] = scheduledExecutorService;
-        }
+        concurrentScheduler = Optional.of(new ConcurrentScheduledExecutor(
+                schedulerCount, threadsPerScheduler, threadNameFormat));
     }
 
     @Override
     protected void doStop()
     {
-        for (int i = 0; i < schedulers.length; i++) {
-            schedulers[i].shutdownNow();
-            schedulers[i] = null;
+        if (concurrentScheduler.isPresent()) {
+            concurrentScheduler.get().shutdownNow();
         }
     }
 
     @Override
     public Task schedule(Runnable task, long delay, TimeUnit unit)
     {
-        ScheduledExecutorService scheduler = schedulers[ThreadLocalRandom.current().nextInt(schedulers.length)];
+        ScheduledExecutorService scheduler = concurrentScheduler.get().getScheduledExecutorService();
         if (scheduler == null) {
```

Comment:
Does this check make sense now? The optional.get will throw no such element if this is invoked before doStart().

Also should getNextScheduledExecutorService() be public? Maybe you should submit tasks and let the service figure out what to do?

Reply:
Good point,  updated.